src/lib.rs

```diff
@@ -85,19 +90,34 @@ impl From<NormalMapFormat> for ispc::NormalMapFormat {
 
 /// Describes a source image which can be used for [`downsample()`]
 /// The pixel data is stored as a slice to avoid unnecessarily cloning it.
-pub struct Image<'a> {
+pub struct Image<'a, F: ImagePixelFormat> {
     pixels: &'a [u8],
     width: u32,
     height: u32,
+    pixel_stride_in_bytes: usize,
+    format: F,
 }
 
-impl<'a> Image<'a> {
+impl<'a, F: ImagePixelFormat> Image<'a, F> {
     /// Creates a new source image from the given pixel data slice, dimensions and format.
-    pub fn new(pixels: &'a [u8], width: u32, height: u32) -> Self {
+    pub fn new(pixels: &'a [u8], width: u32, height: u32, format: F) -> Self {
+        let pixel_size = format.pixel_size();
+        Self::new_with_pixel_stride(pixels, width, height, format, pixel_size)
+    }
+
+    pub fn new_with_pixel_stride(
+        pixels: &'a [u8],
+        width: u32,
+        height: u32,
+        format: F,
+        pixel_stride_in_bytes: usize,
```

Comment:
I guess this is to skip pixels without a meaning, i.e. an `(s)rgb(a)8` loaded PNG where only the `RG` channels are valid for an encoded normal? Could we achieve the same by having the input texture be a plain old `(s)rgb(a)8` format, and the output an `r8g8normal` thing however it is called?

Just thinking out loud what could happen with ambiguity between `Image::format` and `Image::pixel_stride_in_bytes`.